### ISSUE TITLE
Make dagger CLI easy to script

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#3213]
## Overview

In the upcoming *cloak* release (0.3) the primary use of Dagger will be to embed it in a script. In the case of shell scripts (bash, zsh, powershell etc) the mechanism for scripting will be invoking the `dagger` CLI directly.

Therefore it is crucially important that `dagger` be very easy to script. In particular it should be very easy to make multiple calls to `dagger`, and connect their inputs and outputs in a pipeline. Also known as "stitching".

Generally, all command-line tools should be made easy to call in a shell script.

**Comment by marcosnils** [comment]
Challenges I see particularly in Dagger vs scripting any other tools: 

- A single query can run multiple steps in parallel by leveraging graphql field aliases, it's important to document it so users can make the most while using it. 
- Outputs usually require third party tools to parse like `jq`, so it'd be great to help out here.
- Writing queries in a bash script is error prone and clunky as 99% of the times it involves some heredoc + interpolation handling
- Capturing errors and exit codes? i.e: exit code when a query fails vs could not connect to engine

**Comment by shykes** [comment]
> * A single query can run multiple steps in parallel by leveraging graphql field aliases, it's important to document it so users can make the most while using it.

I agree, but is there something we should do about this that is specific to shell scripting?

> * Outputs usually require third party tools to parse like `jq`, so it'd be great to help out here.

I think `dagger` should offer common helpers for input and output formats, since it's a fundamental part of stitching queries together. You should not require an external tool for very common uses of the CLI.

> * Writing queries in a bash script is error prone and clunky as 99% of the times it involves some heredoc + interpolation handling

You're thinking of proposing a queries file on the side? I think that's a viable approach and you should propose more details here :)

> * Capturing errors and exit codes? i.e: exit code when a query fails vs could not connect to engine

I guess we can copy whatever `curl` does here?

**Comment by mircubed** [comment]
@aluzzardi I'm worried about your bandwidth for this one? I will put it in stuck for now since I think this can be a blocker for release if you don't have someone to help you with it.

**Comment by samalba** [comment]
Please take a look at #3268. My experience so far:

- I like relying on `jq` for formatting the output. I prefer this than dagger trying to do any magic, `jq` is far more powerful and easy enough for simple selection.
- The inlined queries does not have issues with escaping thanks to the `--set` cli arg

There is room for improvement (like the session management for example) but it's totally usable with the current CLI and easy enough.

I would close this issue and instead open more scoped one later (for instance: implement the session management, write more guides and examples, etc...).

**Comment by sipsma** [comment]
Relevent issue: https://github.com/dagger/dagger/issues/3082

That was made a long time ago (in the previous cloak repo), but if nothing has changed then there's a problem with the CLI where it can't correctly parse lists of strings w/ `--set`. I.e. you can't really use alpine because it asks for a list of `pkgs`.

**Comment by samalba** [comment]
> That was made a long time ago (in the previous cloak repo), but if nothing has changed then there's a problem with the CLI where it can't correctly parse lists of strings w/ `--set`. I.e. you can't really use alpine because it asks for a list of `pkgs`.

True (although I did not need it), `--set` is only needed when you need to pass info in between queries (output of one query is needed as input for another one).

Given that we mostly need ids (secret, fs), the current support is enough short term. For Alpine, you can simply set your packages in the query itself ([example here](https://github.com/dagger/dagger/blob/6fbd88b0fedbf81e8e81ee13d6ff27c8547956b2/examples/bash/yarn-build.sh#L26)), you don't need `--set` for that, except if one day you dynamically generate your pkg list from another query, which would be kind of weird.
